aptos-move/move-examples/data_structures/sources/smart_table.move

```diff
@@ -0,0 +1,358 @@
+        let i = 0;
+        let j = vector::length(old_bucket);
+        let len = j;
+        while (i < j) {
+            let entry = vector::borrow(old_bucket, i);
+            let index = bucket_index(table.level, table.num_buckets, entry.hash);
+            if (index == new_bucket_index) {
+                j = j - 1;
+                vector::swap(old_bucket, i, j);
+            } else {
+                i = i + 1;
+            };
+        };
+        while (j < len) {
+            let entry = vector::pop_back(old_bucket);
+            vector::push_back(&mut new_bucket, entry);
+            len = len - 1;
+        };
+        table_with_length::add(&mut table.buckets, new_bucket_index, new_bucket);
+    }
+
+    /// Return the expected bucket index to find the hash.
+    fun bucket_index(level: u8, num_buckets: u64, hash: u64): u64 {
+        let index = hash % (1 << (level + 1));
+        if (index < num_buckets) {
+            // in existing bucket
+            index
+        } else {
+            // in unsplitted bucket
+            index % (1 << level)
+        }
+    }
+
+    /// Acquire an immutable reference to the value which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    public fun borrow<K: drop, V>(table: &SmartTable<K, V>, key: K): &V {
+        let index = bucket_index(table.level, table.num_buckets, sip_hash_from_value(&key));
+        let bucket = table_with_length::borrow(&table.buckets, index);
+        let i = 0;
+        let len = vector::length(bucket);
+        while (i < len) {
+            let entry = vector::borrow(bucket, i);
+            if (&entry.key == &key) {
+                return &entry.value
+            };
+            i = i + 1;
+        };
+        abort error::invalid_argument(ENOT_FOUND)
+    }
+
+    /// Acquire a mutable reference to the value which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    public fun borrow_mut<K: drop, V>(table: &mut SmartTable<K, V>, key: K): &mut V {
+        let index = bucket_index(table.level, table.num_buckets, sip_hash_from_value(&key));
+        let bucket = table_with_length::borrow_mut(&mut table.buckets, index);
+        let i = 0;
+        let len = vector::length(bucket);
+        while (i < len) {
+            let entry = vector::borrow_mut(bucket, i);
+            if (&entry.key == &key) {
+                return &mut entry.value
+            };
+            i = i + 1;
+        };
+        abort error::invalid_argument(ENOT_FOUND)
+    }
+
+    /// Returns true iff `table` contains an entry for `key`.
+    public fun contains<K: drop, V>(table: &SmartTable<K, V>, key: K): bool {
+        let index = bucket_index(table.level, table.num_buckets, sip_hash_from_value(&key));
+        let bucket = table_with_length::borrow(&table.buckets, index);
+        let i = 0;
+        let len = vector::length(bucket);
+        while (i < len) {
```

Comment:
Here and in other places. You should use vector::any: https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/move-stdlib/sources/vector.move#L221. Here and in other places. Might also be better if this is done in a separate function since this code seems to be used in many functions here.